pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -489,71 +491,121 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
-					}
-				}
-
-				return true, nil
-			})
-			if err != nil {
-				return false, err
-			}
-		}
 		return true, nil
 	})
 	if err != nil {
 		return err
 	}
 
-	// Anything left in addrs is a new member that needs to be added to a pool
-	for addr := range addrs {
-		_, err := v2_pools.CreateAssociateMember(lbaas.network, poolID, v2_pools.MemberCreateOpts{
-			Address:      addr,
-			ProtocolPort: int(ports[0].NodePort),
-			SubnetID:     lbaas.opts.SubnetId,
-		}).Extract()
+	// Compose Set of member (addresses) that _should_ exist
+	addrs := map[string]bool{}
+	for _, host := range hosts {
+		addr, err := getAddressByName(lbaas.compute, host)
 		if err != nil {
 			return err
 		}
-		waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+		addrs[addr] = true
 	}
 
+	// Remove members that are no longer required
+	for _, pool := range poolsLists {
+		err := v2_pools.ListAssociateMembers(lbaas.network, pool.ID, v2_pools.MemberListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+			membersList, err := v2_pools.ExtractMembers(page)
+			if err != nil {
+				return false, err
+			}
+			for _, member := range membersList {
+				if _, found := addrs[member.Address]; found {
+					// Member already exists, remove from update list
+					addrs[member.Address] = false
+				} else {
+					// Member needs to be deleted
+					waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+					err = v2_pools.DeleteMember(lbaas.network, pool.ID, member.ID).ExtractErr()
+					if err != nil {
+						return false, err
+					}
+					waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+				}
+			}
+			return true, nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Add new members for each port in corresponding listener/pool
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+	for _, port := range ports {
+		// Find appropriate listener for this port
+		var listenerID string
+		for _, listener := range listenersLists {
+			if listener.ProtocolPort == int(port.Port) && listener.Protocol == string(port.Protocol) {
+				listenerID = listener.ID
+				break
+			}
+		}
+		if listenerID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required listener for port %d and protocol %s", loadBalancerName, port.Port, port.Protocol)
+		}
+
+		// Find appropriate pool for found listener
+		var poolID string
+		for _, pool := range poolsLists {
+			for _, currentListener := range pool.Listeners {
+				if listenerID == currentListener.ID {
+					poolID = pool.ID
+					break
+				}
+			}
+			if poolID != "" {
+				break
+			}
+		}
+		if poolID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required pool for listener %s", loadBalancerName, listenerID)
+		}
+
+		// Create a member for each new host
+		for addr := range addrs {
+			if !addrs[addr] {
+				// skip existing member
+				continue
```

Comment:
Thank you for explaining with PR.  I like the adding new members for each port before removing not-required members.  Through testing in OS, the first solution still worked as intended even if a failure happened while looking up existing members since state on members that have been removed/added is not kept.  I will take your recommendation for keeping adding/removing members in the same overall port loop.